presto-orc/src/main/java/com/facebook/presto/orc/reader/TimestampSelectiveStreamReader.java

```diff
@@ -0,0 +1,430 @@
+                if (filter.testLong(value)) {
+                    if (outputRequired) {
+                        values[outputPositionCount] = value;
+                        if (nullsAllowed && presentStream != null) {
+                            nulls[outputPositionCount] = false;
+                        }
+                    }
+                    outputPositions[outputPositionCount] = position;
+                    outputPositionCount++;
+                }
+            }
+            streamPosition++;
+        }
+        return streamPosition;
+    }
+
+    private int readAllNulls(int[] positions, int positionCount)
+            throws IOException
+    {
+        presentStream.skip(positions[positionCount - 1]);
+
+        if (nullsAllowed) {
+            outputPositionCount = positionCount;
+            allNulls = true;
+        }
+        else {
+            outputPositionCount = 0;
+        }
+
+        return positions[positionCount - 1] + 1;
+    }
+
+    private int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        // filter == null implies outputRequired == true
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                nulls[i] = true;
+            }
+            else {
+                values[i] = decodeTimestamp(secondsStream.next(), nanosStream.next(), baseTimestampInSeconds);
+                if (presentStream != null) {
+                    nulls[i] = false;
+                }
+            }
+            streamPosition++;
+        }
+        outputPositionCount = positionCount;
+        return streamPosition;
+    }
+
+    private void skip(int items)
+            throws IOException
+    {
+        if (secondsStream == null && nanosStream == null) {
+            presentStream.skip(items);
+        }
+        else if (presentStream != null) {
+            int dataToSkip = presentStream.countBitsSet(items);
+            secondsStream.skip(dataToSkip);
+            nanosStream.skip(dataToSkip);
+        }
+        else {
+            secondsStream.skip(items);
+            nanosStream.skip(items);
+        }
+    }
+
+    private void ensureValuesCapacity(int capacity, boolean recordNulls)
+    {
+        if (values == null || values.length < capacity) {
+            values = new long[capacity];
+        }
+
+        if (recordNulls) {
+            if (nulls == null || nulls.length < capacity) {
+                nulls = new boolean[capacity];
+            }
+        }
+    }
+
+    private void ensureOutputPositionsCapacity(int capacity)
+    {
+        if (outputPositions == null || outputPositions.length < capacity) {
+            outputPositions = new int[capacity];
```

Comment:
Could you use the new Arrays.ensureCapacity methods here and in ensureValuesCapacity?